crates/sui-core/src/authority.rs

```diff
@@ -1415,42 +1415,112 @@ impl AuthorityState {
 
         let execution_start_time = Instant::now();
 
-        let input_objects = self.read_objects_for_execution(
-            tx_guard.as_lock_guard(),
-            certificate,
-            execution_env.assigned_versions,
-            epoch_store,
-        )?;
-
-        if execution_env.expected_effects_digest.is_none() {
-            // We could be re-executing a previously executed but uncommitted transaction, perhaps after
-            // restarting with a new binary. In this situation, if we have published an effects signature,
-            // we must be sure not to equivocate.
-            // TODO: read from cache instead of DB
-            execution_env.expected_effects_digest =
-                epoch_store.get_signed_effects_digest(tx_digest)?;
+        let execution_guard = self.execution_lock_for_executable_transaction(certificate);
+        // Any caller that verifies the signatures on the certificate will have already checked the
+        // epoch. But paths that don't verify sigs (e.g. execution from checkpoint, reading from db)
+        // present the possibility of an epoch mismatch. If this cert is not finalzied in previous
+        // epoch, then it's invalid.
```

Comment:
Maybe `let execution_guard = self.execution_lock_for_executable_transaction(certificate);` should be under this comment too, and mention taking execution_guard does the checks mentioned in the comment?